sdk/core/Azure.Core/src/Shared/GeoRedundantFallbackPolicy.cs

```diff
@@ -81,72 +79,97 @@ public override void OnSendingRequest(HttpMessage message)
                 return;
 
             bool isRead = message.Request.Method == RequestMethod.Get || message.Request.Method == RequestMethod.Head;
+            Fallback fallback = isRead ? _readFallback : _writeFallback;
 
-            ref int fallbackIndex = ref isRead ? ref _readFallbackIndex : ref _writeFallbackIndex;
-            ref long fallbackTicks = ref isRead ? ref _readFallBackTicks : ref _writeFallbackTicks;
-            string[] fallbackHosts = isRead ? _readFallbackHosts : _writeFallbackHosts;
+            if (fallback.Hosts.Length == 0)
+                return;
 
             if (message.ProcessingContext.RetryNumber == 0)
             {
+                // store the primary host in the message
                 SetPrimaryHost(message);
                 // set the host based on the fallback information
-                UpdateHost(message, fallbackHosts, ref fallbackIndex, ref fallbackTicks);
+                UpdateHostIfNeeded(message, fallback);
                 return;
             }
 
-            if (fallbackHosts.Length == 0)
-                return;
-
-            int current = Volatile.Read(ref fallbackIndex);
-            long currentTicks = Volatile.Read(ref fallbackTicks);
-
-            // we should only advance if another thread hasn't already done so
-            if ((current == -1 && message.Request.Uri.Host.Equals(GetPrimaryHost(message), StringComparison.Ordinal)) ||
-                (current != -1 && message.Request.Uri.Host.Equals(fallbackHosts[current], StringComparison.Ordinal)))
-            {
-                int next = current + 1;
+            fallback.AdvanceIfNeeded(message);
+            UpdateHostIfNeeded(message, fallback);
+        }
 
-                // reset to null to indicate primary should be used when we reach the end
-                if (next >= fallbackHosts.Length)
-                    next = -1;
+        private static void UpdateHostIfNeeded(HttpMessage message, Fallback fallback)
```

Comment:
could this also go inside `Fallback`?

Reply:
It could, but I wanted to only include methods in Fallback when mutating the fallback as opposed to mutating the message itself.